### DISCUSSION TITLE
mpb in meep with dispersive materials

### DISCUSSION THREAD

**Discussion by ryouta1120** [NanoComp/meep#1972] [Q&A]
I'm interested in photonic crystals, so I'm considering using mpb packaged in meep.
Here, I have a simple question. 
Is it available to apply dispersive Drude-Lorentz metals on mpb?

According to the discription of [mpb](https://mpb.readthedocs.io/en/latest/), I guess it only allows to specify as constant.
However, testing some materials in materials.py, I could run mpb.modesolver without any error.
My environment is built by anaconda.

**Comment by oskooi** [comment]
> Is it available to apply dispersive Drude-Lorentz metals on mpb?

No. As stated in the "Features" list of https://mpb.readthedocs.io/en/latest/:

```
Support for arbitrary, anisotropic dielectrics including gyrotropic/magneto-optic materials 
and non-orthogonal unit cells. Lossy and wavelength-dependent ε and μ are not supported.
```
Note, however, that Meep (not MPB) has a frequency-domain eigensolver which you can use for structures involving metals: https://meep.readthedocs.io/en/latest/Eigensolver_Math/. This feature has not yet been thoroughly tested and so convergence may be an issue.

**Reply by ryouta1120** [reply]
Thank you for replying.

Now in my case, is it the best way to use Hatmienv following [band diagram tutorial](https://meep.readthedocs.io/en/latest/Python_Tutorials/Resonant_Modes_and_Transmission_in_a_Waveguide_Cavity/)?

**Reply by smartalecH** [reply]
You could also try the frequency-domain shift-invert eigensolver within Meep. The manual description is [here](https://meep.readthedocs.io/en/latest/Eigensolver_Math/). There's an example script [here](https://github.com/NanoComp/meep/blob/d662b0c4f518a4d14ad530394dd2d10d82887652/python/tests/test_eigfreq.py).

You can use PML, Lorenz-Drude materials, etc.

If you just care about generating band diagrams, however, Harminv is probably the way to go. It will be significantly faster and more efficient.

**Reply by ryouta1120** [reply]
Thank you for your suggestion.

I will try both.
